Document secure Kserve authentication via automated tests

madmecodes opened the pull request:
# Pull Request Template for Kubeflow Manifests

## ✏️ Summary of Changes
This PR addresses a security issue with KServe inference endpoints by applying oauth2-proxy authentication to the cluster-local-gateway. Currently, KServe endpoints lack proper authentication because they use the cluster-local-gateway which has an `ALLOW` policy instead of proper oauth2-proxy authentication.

Changes include:
1. Modified the cluster-local-gateway AuthorizationPolicy to use oauth2-proxy authentication instead of allowing all traffic without authentication
2. Updated the KServe test workflow to:
   - Use proper Kubeflow user namespaces instead of generic test namespaces
   - Test with tokens from the appropriate service accounts
   - Apply predictor-specific AuthorizationPolicy for tests

This ensures KServe inference endpoints are properly secured and require authentication, just like other Kubeflow services accessed through the istio-ingressgateway.

## 📦 Dependencies
No direct dependencies on other PRs. This PR combines improvements that were planned for PR #2936 with the security fix.

## 🐛 Related Issues
- Fixes #2811 (Secure KServe inferencing with proper authentication)
- Related to #2936 (Update KServe tests with real Kubeflow namespaces)
- Helps unblock #2907 (ISTIO CNI by default)

## ✅ Contributor Checklist
  - [x] I have tested these changes with kustomize. See [Installation Prerequisites](https://github.com/kubeflow/manifests#prerequisites).
  - [x] All commits are [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits) to satisfy the DCO check.

---     
 
> You can join the CNCF Slack and access our meetings at the [Kubeflow Community](https://www.kubeflow.org/docs/about/community/) website. Our channel on the CNCF Slack is here [**#kubeflow-platform**](https://app.slack.com/client/T08PSQ7BQ/C073W572LA2).

madmecodes commented:
@juliusvonkohout looking for the feedback

juliusvonkohout commented:
/retest

madmecodes commented:
> /retest
<img width="714" alt="Screenshot 2025-03-18 at 11 05 09 PM" src="https://github.com/user-attachments/assets/0af9ef9d-41a4-4834-9545-a4ce256194e8" />


namespace creation is failing, i am debugging that, i will tag u once the issue is solved

juliusvonkohout commented:
It works differently. Please wait until https://github.com/kubeflow/manifests/pull/3062 is merged, then rebase. There is no need for you to debug namespace creation.

juliusvonkohout commented:
You can also just create a new branch based on https://github.com/kubeflow/manifests/tree/merge-the-kserve-tests.

juliusvonkohout commented:
https://github.com/kubeflow/manifests/pull/3062 is merged, so you can just start from master again.

juliusvonkohout commented on .github/workflows/kserve_cni_test.yaml:
this file will be deleted if you rebase

juliusvonkohout commented on .github/workflows/kserve_test.yaml:
this file weill be deleted when you rebase to master

juliusvonkohout commented on .github/workflows/kserve_m2m_test.yaml:
- name: Create kubeflow namespace
      run: kustomize build common/kubeflow-namespace/base | kubectl apply -f -

can be removed because it is a duplicate of 

    - name: Create KF Profile
      run: kustomize build common/user-namespace/base | kubectl apply -f -

It helps if 

    - name: Create KF Profile
      run: kustomize build common/user-namespace/base | kubectl apply -f -

comes before 

    - name: Install KF Multi Tenancy
      run: ./tests/gh-actions/install_multi_tenancy.sh

as in https://github.com/kubeflow/manifests/blob/88801f6ce84ac7eab3e2883dd3d80cb63ccb6ab3/.github/workflows/pipeline_run_from_notebook.yaml#L45C1-L46C55

juliusvonkohout commented on .github/workflows/kserve_m2m_test.yaml:
```diff
@@ -44,11 +43,46 @@ jobs:
+      run: ./tests/gh-actions/install_multi_tenancy.sh
+
+    - name: Install kubeflow-istio-resources
+      run: kustomize build common/istio-1-24/kubeflow-istio-resources/base | kubectl apply -f -
```
also this

```
    - name: Install kubeflow-istio-resources
      run: kustomize build common/istio-1-24/kubeflow-istio-resources/base | kubectl apply -f -
```
should be higher up as in https://github.com/kubeflow/manifests/blob/88801f6ce84ac7eab3e2883dd3d80cb63ccb6ab3/.github/workflows/pipeline_run_from_notebook.yaml#L39

juliusvonkohout commented on .github/workflows/kserve_m2m_test.yaml:
```diff
@@ -44,11 +43,46 @@ jobs:
+    - name: Create KF Profile
+      run: kustomize build common/user-namespace/base | kubectl apply -f -
+
+    - name: Wait for KF Profile to be ready
```
this can be removed if you rebase and address the comments above

madmecodes commented:
<img width="824" alt="Screenshot 2025-03-21 at 12 22 56 PM" src="https://github.com/user-attachments/assets/b6ab17e9-5b7e-4fd9-b056-0d390e078464" />

is it that, params.env isn't loading?
<img width="531" alt="Screenshot 2025-03-21 at 12 24 01 PM" src="https://github.com/user-attachments/assets/6b62fb73-b46e-4f44-90be-f77475db07c5" />

juliusvonkohout commented:
@madmecodes now we are progressed until ` FAILED test_sklearn.py::test_sklearn_kserve - requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: http://localhost:8080/v1/models/isvc-sklearn:predict`

juliusvonkohout commented:
```
 test_sklearn.py:58: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
utils.py:67: in predict
    return predict_str(
utils.py:125: in predict_str
    response.raise_for_status()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Response [403]>

    def raise_for_status(self):
        """Raises :class:`HTTPError`, if one occurred."""
    
        http_error_msg = ""
        if isinstance(self.reason, bytes):
            # We attempt to decode utf-8 first because some servers
            # choose to localize their reason strings. If the string
            # isn't utf-8, we fall back to iso-8859-1 for all other
            # encodings. (See PR #3538)
            try:
                reason = self.reason.decode("utf-8")
            except UnicodeDecodeError:
                reason = self.reason.decode("iso-8859-1")
        else:
            reason = self.reason
    
        if 400 <= self.status_code < 500:
            http_error_msg = (
                f"{self.status_code} Client Error: {reason} for url: {self.url}"
            )
    
        elif 500 <= self.status_code < 600:
            http_error_msg = (
                f"{self.status_code} Server Error: {reason} for url: {self.url}"
            )
    
        if http_error_msg:
>           raise HTTPError(http_error_msg, response=self)
E           requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: http://localhost:8080/v1/models/isvc-sklearn:predict

/opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages/requests/models.py:1024: HTTPError
------------------------------ Captured log call -------------------------------
INFO     root:utils.py:103 M2M Token Found.
INFO     root:utils.py:114 Sending Header = {'Host': 'isvc-sklearn.kubeflow-user-example-com.example.com', 'Content-Type': 'application/json', 'Authorization': '***'}
INFO     root:utils.py:115 Sending url = http://localhost:8080/v1/models/isvc-sklearn:predict
INFO     root:utils.py:116 Sending request data: {"instances": [[6.8, 2.8, 4.8, 1.4], [6.0, 3.4, 4.5, 1.6]]}
INFO     root:utils.py:118 Got response code 403, content b'RBAC: access denied'
=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: Type google._upb._message.MessageMapContainer uses PyType_Spec with a metaclass that has custom tp_new. This is deprecated and will no longer be allowed in Python 3.14.

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: Type google._upb._message.ScalarMapContainer uses PyType_Spec with a metaclass that has custom tp_new. This is deprecated and will no longer be allowed in Python 3.14.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED test_sklearn.py::test_sklearn_kserve - requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: http://localhost:8080/v1/models/isvc-sklearn:predict
================== 1 failed, 2 warnings in 112.54s (0:01:52) ===================
Error: Process completed with exit code 1.
```

juliusvonkohout commented:
``` 
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-in-cluster-kserve
spec:
  rules:
    - to:
        - operation:
            paths:
              - /v1/models/*
              - /v2/models/*
```

could be added and adapted to only allow traffic from the same namespace or istio-system

Path based routing is also an option

```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: ${INFERENCESERVICE}-external
spec:
  gateways:
    - kubeflow/kubeflow-gateway # migh have to be updated
  hosts:
    - '*'
  http:
    - match:
        - uri:
            prefix: /kserve/${NAMESPACE}/${INFERENCESERVICE}/
      rewrite:
        uri: /
      route:
        - destination:
            host: knative-local-gateway.istio-system.svc.cluster.local # might have to be updated
          headers:
            request:
              set:
                Host: ${INFERENCESERVICE}-predictor-default.${NAMESPACE}.svc.cluster.local
          weight: 100
      timeout: 300s
```

juliusvonkohout commented:
https://github.com/kubeflow/kubeflow/pull/6627 is also interesting

madmecodes commented:
@juliusvonkohout 

i shall now make PR compatible with istio-cni also right? its currently setup with regular istio (init containers)

What i can understand right now:

**Before PR: Regular Istio (non-CNI)**

1. KServe endpoints were secured using the cluster-local-gateway with an ALLOW policy
2. This didn't properly integrate with Kubeflow's authentication system

**After PR: Still using regular Istio**

1. Changed the cluster-local-gateway to use oauth2-proxy authentication
2. Added specific AuthorizationPolicies for KServe endpoints
3. This properly secures KServe with the same authentication mechanism as other Kubeflow services

**but now,**
Failing Dex workflow: Using Istio CNI
Which is part of a migration effort toward using CNI as the default (PR #3061)

<img width="500" alt="Screenshot 2025-03-23 at 11 24 05 AM" src="https://github.com/user-attachments/assets/b5757cbe-0678-43ef-bb09-669958ce77f4" />

juliusvonkohout commented:
Hello,

1. the dex test is just a bit flaky. Lets try to make it more robust in this PR as well. 
2. Alright let me enable istio-cni by default.
3. do not hardcode "kubeflow-user-example-com" in the authorizationpolicy, but make it pick up the namespace the authorizationpolicy is deployed in.

juliusvonkohout commented:
@kromanow94 can you review as well?

juliusvonkohout commented on common/istio-1-24/cluster-local-gateway/base/gateway-authorizationpolicy.yaml:
@akagami-harsh can you cross check whether this file is auto generated?

akagami-harsh replied:
The `gateway-authorizationpolicy.yaml` file appears to be a manually created file that doesn't change when running the `synchronize-istio-manifests.sh` script.

madmecodes commented:
@juliusvonkohout  Currently Auth Policy is allowing access to models, via any namespace, if it has a valid JWT we need to change it to so that only specified namespaces in AuthPolicy can access them, am i thinking in the right direction?

i am testing it locally, when its too strict, it forbids all if too permissive, allows all (with valid JWT), working on a middle path, am i working on right problem?
<img width="500" alt="Screenshot 2025-03-24 at 7 32 24 PM" src="https://github.com/user-attachments/assets/44a3f01f-92b2-4a22-81ab-02e7b7a299f3" />

juliusvonkohout commented:
> @juliusvonkohout Currently Auth Policy is allowing access to models, via any namespace, if it has a valid JWT we need to change it to so that only specified namespaces in AuthPolicy can access them, am i thinking in the right direction?
> 
> i am testing it locally, when its too strict, it forbids all if too permissive, allows all (with valid JWT), working on a middle path, am i working on right problem? <img alt="Screenshot 2025-03-24 at 7 32 24 PM" width="500" src="https://private-user-images.githubusercontent.com/137040550/426112167-44a3f01f-92b2-4a22-81ab-02e7b7a299f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4Mjg4MjMsIm5iZiI6MTc0MjgyODUyMywicGF0aCI6Ii8xMzcwNDA1NTAvNDI2MTEyMTY3LTQ0YTNmMDFmLTkyYjItNGEyMi04MWFiLTAyZTdiN2EyOTlmMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNFQxNTAyMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZThkNDQzYTkxY2Y4MTMxODg5OTQ3NTJhMmZhMjkwMjZjNTFjNmRkNjg5Y2MyZTliYjFkMjlmOTJlZmRiZjFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Qc1-IC_PKqwSKPkaph903Qla92wWs_ajt0sW50EPih8">
> <img alt="Screenshot 2025-03-24 at 7 36 45 PM" width="500" src="https://private-user-images.githubusercontent.com/137040550/426114133-902fb042-70f5-4bba-a094-0875161365da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4Mjg4MjMsIm5iZiI6MTc0MjgyODUyMywicGF0aCI6Ii8xMzcwNDA1NTAvNDI2MTE0MTMzLTkwMmZiMDQyLTcwZjUtNGJiYS1hMDk0LTA4NzUxNjEzNjVkYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNFQxNTAyMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMjNhNzMzNDdlODIxY2JjZDEyOTUyNDYzMzBiMDA5OTFkODJhNWJlYmYxZThiZmJiZjgwMWY0ZTMxMjcxZjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DZcJ_ys3H37qltJLN_R65GhZJTc4xp6eOLBS5hWnkJQ"> <img alt="Screenshot 2025-03-24 at 7 36 58 PM" width="500" src="https://private-user-images.githubusercontent.com/137040550/426114154-7d661953-7fed-4dc9-b1de-3f62cc714f21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4Mjg4MjMsIm5iZiI6MTc0MjgyODUyMywicGF0aCI6Ii8xMzcwNDA1NTAvNDI2MTE0MTU0LTdkNjYxOTUzLTdmZWQtNGRjOS1iMWRlLTNmNjJjYzcxNGYyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNFQxNTAyMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYzAzNGZhMTU1MjczNjQxOGM1ZjliMDRiMWNiMTFhZjIwMWU1OWNhNzAyMTZhYWVlZmQxZWE0OTlkMTllOTcyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZmyaGanJvYpDqRNDAs5L-gnsYeMEN5jzJK8Z13zNfqA">

Yes, that is what we want. Only mebers/serviceaccounts of the same namespace should be able to access it. Maybe you can work with the default-editor, default and default-viewer serviceaccount of the same namespace as principals in the authorizationpolicy.

juliusvonkohout commented:
Let me do some changes, merge the PR and then you can create a follow-up PR @madmecodes

juliusvonkohout commented:
/lgtm
/approve

juliusvonkohout commented:
Please make sure that you rebase to master before you create the new branch / PR.

juliusvonkohout commented:
In the follow up PR you can either remove or secure the cluster-local-gateway. Whatever works for kserve and keeps all  security tests green.
